apps/extension/client/src/checkTypegenNesting.ts

```diff
@@ -0,0 +1,150 @@
+    ? `${existingTsPattern}, ${typeGenPattern}`
+    : typeGenPattern;
+
+  // Update file nesting patterns with all existing patterns and our updated pattern
+  fileNestingConfig.update(
+    "patterns",
+    {
+      ...fileNestingPatterns, // Reuse existing patterns
+      [typeGenPatternKey]: updatedPattern, // Add our updated pattern
+    },
+    true
+  );
+};
+
+const disableTypeGenNesting = (
+  xstateConfig: WorkspaceConfiguration,
+  fileNestingConfig: WorkspaceConfiguration,
+  fileNestingPatterns: object
+) => {
+  xstateConfig.update("nestTypegenFiles", false, true);
+
+  // Check if the user has a pattern defined for ts files
+  const existingTsPattern = fileNestingPatterns[typeGenPatternKey];
+
+  // If there's no patterns defined for ts files we don't need to do anything
+  if (existingTsPattern === "") return;
+
+  // Remove our typegen pattern from the existing patterns
+  const updatedPattern = existingTsPattern
+    .split(",")
+    .map((part) => part.trim())
+    .filter((part) => part !== typeGenPattern)
+    .join(",");
+
+  // Update file nesting patterns with all existing patterns and our updated pattern
+  fileNestingConfig.update(
+    "patterns",
+    {
+      ...fileNestingPatterns, // Reuse existing patterns
+      [typeGenPatternKey]: updatedPattern, // Add our updated pattern
+    },
+    true
+  );
+};
+
+const addConfigChangeHandler = () => {
+  vscode.workspace.onDidChangeConfiguration((event) => {
+    if (event.affectsConfiguration("xstate.nestTypegenFiles")) {
+      // VSCode's configuration for file nesting
+      const fileNestingConfig = vscode.workspace.getConfiguration(
+        "explorer.fileNesting"
+      );
+
+      const xstateConfig = vscode.workspace.getConfiguration("xstate");
+      const updatedNestTypegenFiles =
+        xstateConfig.get<boolean>("nestTypegenFiles");
+
+      if (updatedNestTypegenFiles) {
+        enableTypeGenNesting(
+          fileNestingConfig,
+          fileNestingConfig.get("patterns")
+        );
+      } else {
+        disableTypeGenNesting(
+          xstateConfig,
+          fileNestingConfig,
+          fileNestingConfig.get("patterns")
+        );
+      }
+    }
+  });
+};
+
+/*
+ * This checks that the users' config for nesting typegen files is set up correctly.
+ * We try to make it easy for the user to enable file nesting by showing a prompt if they want file nesting but haven't defined a pattern for ts files yet.
+ */
+export const handleTypegenNestingConfig = () => {
+  // VSCode's configuration for file nesting
+  const fileNestingConfig = vscode.workspace.getConfiguration(
+    "explorer.fileNesting"
+  );
+
+  // If there's no fileNestingConfig with patterns then we can't check the nesting, maybe the user is on a version of VSCode before 1.67
+  if (!fileNestingConfig || !fileNestingConfig.has("patterns")) return;
+
+  // VSCode's configuration for xstate
+  const fileNestingPatterns = fileNestingConfig.get<object>("patterns");
```

Comment:
How safe this generic param is? Can I do something stupid like editing `patterns` manually to a `null` or smth and blow up our extension this way? :P Or is there some kind of a JSON schema applied on VScode's side that prevents such errors?

Reply:
VSCode does have a schema, but it's not enforced, so I added some type guards. However, we can't guard completely, so if a user does `"explorer.fileNesting.patterns": null` in their settings, we will get an error because this call `vscode.workspace.getConfiguration("explorer.fileNesting")` will fail.